### PR TITLE
Adaptative classification trainer

### PR DESCRIPTION
**Pull request by MateoLostanlen**
The classification trainer only works for classifications with 5 or more classes. The purpose of this PR is to make it adaptive

### CODE REVIEW

**Review by frgfm — CHANGES_REQUESTED** [review]
Thanks for the PR! While I agree that there would be some errors for multi-class classification with fewer than 5 classes, your PR also infers on the way to produce the output probability, which might be an issue. I suggest keeping this PR about the first aspect (the second would be a good occasion to make a `BinaryClassificationTrainer` class for instance)

Also, could you add a short unittest about this please? :pray:

**Comment by MateoLostanlen** [comment]
Ok I understand your point I will remove the case of binary classification. Yes sorry the test is comming :)

File: holocron/trainer/core.py [hunk 1]
```diff
@@ -348,10 +348,14 @@ def evaluate(self) -> Dict[str, float]:
...
-            pred = out.topk(5, dim=1)[1]
+            pred = out.topk(1, dim=1)[1]
             correct = pred.eq(target.view(-1, 1).expand_as(pred))
             top1 += correct[:, 0].sum().item()
-            top5 += correct.any(dim=1).sum().item()
+
+            if out.shape[1] >= 5:  # Top5 score if there is 5 classes or more
+                pred = out.topk(5, dim=1)[1]
+                top5 += correct.any(dim=1).sum().item()
+
```
Comment (frgfm) [inline holocron/trainer/core.py:358]: the `.topk` would be computed twice here. I would suggest the following:
```
pred = out.topk(5, dim=1)[1] if out.shape[1] >= 5 else out.argmax(dim=1, keepdim=True)
correct = pred.eq(target.view(-1, 1).expand_as(pred))
top1 += correct[:, 0].sum().item()
if out.shape[1] >= 5:
    top5 += correct.any(dim=1).sum().item()
```

File: holocron/trainer/core.py [hunk 2]
```diff
@@ -339,7 +339,7 @@ def evaluate(self) -> Dict[str, float]:
 
         self.model.eval()
 
-        val_loss, top1, top5, num_samples = 0., 0, 0, 0
+        val_loss, top1, top5, num_samples = 0., 0., 0., 0
```
Comment (frgfm) [inline holocron/trainer/core.py:342]: I'm curious: was there some necessary casting here? I thought in python 3, the division of integer was casted to float automatically

File: test/test_trainer.py [hunk 3]
```diff
@@ -132,6 +132,37 @@ def test_classification_trainer(self):
...
+        model = nn.Sequential(nn.Conv2d(3, 32, 3), nn.ReLU(inplace=True),
+                              GlobalAvgPool2d(flatten=True), nn.Linear(32, 6))
+        train_loader = torch.utils.data.DataLoader(MockClassificationDataset(num_it * batch_size),
+                                                   batch_size=batch_size)
+        optimizer = torch.optim.Adam(model.parameters())
+        criterion = torch.nn.CrossEntropyLoss()
+        learner = trainer.ClassificationTrainer(model, train_loader, train_loader, criterion, optimizer)
+
+        learner.evaluate()
+
```
Comment (frgfm) [inline test/test_trainer.py:165]:
```
def _test_classification_trainer(self, num_classes):

num_it = 10
batch_size = 8
# Generate all dependencies
model = nn.Sequential(nn.Conv2d(3, 32, 3), nn.ReLU(inplace=True),
                      GlobalAvgPool2d(flatten=True), nn.Linear(32, num_classes))
train_loader = torch.utils.data.DataLoader(MockClassificationDataset(num_it * batch_size),
                                           batch_size=batch_size)
optimizer = torch.optim.Adam(model.parameters())
criterion = torch.nn.CrossEntropyLoss()
learner = trainer.ClassificationTrainer(model, train_loader, train_loader, criterion, optimizer)

learner.evaluate()
if num_classes < 5:
    self.assertEqual(res['acc5'], 0)

```

and then

```
def test_classification_trainer(self):
    # Fewer than 5 classes
    self._test_classification_trainer(6)
    # More than 5 classes
    self._test_classification_trainer(3)
```
would save some code :)

**Review by frgfm — CHANGES_REQUESTED** [review]
Sorry to bother, I still have a few suggestions :sweat_smile:

File: holocron/trainer/core.py (hunk 2 above)
Comment (MateoLostanlen) [reply holocron/trainer/core.py:342]: Yes agree, I don't remeber why I made this change ..

**Comment by MateoLostanlen** [comment]
> 
> 
> Sorry to bother, I still have a few suggestions 😅

I like suggestions, it makes me learn :) I made the changes, tell me if it's good for you ?

**Review by frgfm — APPROVED** [review]
Thanks for the edits!

**Pull request merged**